src/modules/Game.class.js

```diff
@@ -50,19 +138,168 @@ class Game {
+
+      this.#state[empties[idxSorted][0]][empties[idxSorted][1]] =
+        Math.floor(Math.random() * 10) < 9 ? 2 : 4;
+    }
+  }
+
+  // Matrix transformation transpose operation.
+  #tranpose() {
+    let tempRetriveVar;
+
+    for (let ln = 0; ln < 4; ln++) {
+      for (let cl = ln; cl < 4; cl++) {
+        tempRetriveVar = this.#state[ln][cl];
+        this.#state[ln][cl] = this.#state[cl][ln];
+        this.#state[cl][ln] = tempRetriveVar;
+      }
+    }
+  }
+
+  // Revert each line of this.#state, dont use to revert any array. 
+  #reverse() {
+    this.#state.forEach((e, i, arr) => arr[i].reverse());
+  }
+
+  // Check if movements isn't possible than change this.status.
+  #checkIfLose() {
+    let allSides = [];
+
+    for (let column = 0; column < this.#SIZE; column++) {
+      for (let line = 0; line < this.#SIZE; line++) {
+        if (this.#state[line][column] !== 0) {
+          allSides.push(
+            this.#state[line - 1] !== undefined
+              ? this.#state[line - 1][column]
+              : undefined,
+          );
+
+          allSides.push(
+            this.#state[line + 1] !== undefined
+              ? this.#state[line + 1][column]
+              : undefined,
+          );
+          allSides.push(this.#state[line][column + 1]);
+          allSides.push(this.#state[line][column - 1]);
+
+          if (
+            allSides.includes(0) ||
+            allSides.includes(this.#state[line][column])
+          ) {
+            return;
+          }
+        }
+        allSides = [];
+      }
+    }
+
+    this.#status = 'lose';
+  }
 
   /**
-   * Resets the game.
+   * This method copy two dimensional array.
+   * In this code is always used to copy the state Two dimensional Array
+   *
+   * @param {Array} arr
+   * @returns
    */
-  restart() {}
+  #getStateCopy(arr = this.#state) {
+    return arr.map((e) => e.slice());
+  }
+
+  // Make left movement
+  #makeMove() {
+    let limit;
 
-  // Add your own methods here
+    for (let line = 0; line < 4; line++) {
+      limit = this.#SIZE - 1;
+
+      for (let column = 0; column < limit; column++) {
+        if (this.#state[line][column] === 0) {
+          for (let idx = column; idx < this.#SIZE - 1; idx++) {
+            this.#state[line][idx] = this.#state[line][idx + 1];
+            this.#state[line][idx + 1] = 0;
+          }
+
+          column--;
+          limit--;
+
+          continue;
+        } else if (this.#state[line][column + 1] === 0) {
+          for (let idx = column + 1; idx < this.#SIZE - 1; idx++) {
+            this.#state[line][idx] = this.#state[line][idx + 1];
+            this.#state[line][idx + 1] = 0;
+          }
+
+          limit--;
+          column--;
+
+          continue;
+        } else if (
+          this.#state[line][column] === this.#state[line][column + 1]
+        ) {
+          this.#state[line][column] *= 2;
+          this.#state[line][column + 1] = 0;
```

Comment:
Instead of adding a 0 that will be removed in the next iteration handle it here

Reply:
This zero is because I need to change a not zero value to zero because I have a sum. The case if this zero need to removed( acctually replaced) or not is checked in next step of loop. Can you clarify if is necessary additional code?